r1cs/gadgets/crypto/src/prf/ripemd160.rs

```diff
@@ -0,0 +1,458 @@
+                    Boolean::xor(cs.ns(|| format!("A XOR B XOR C {}", i)), &result,&c)
+                }
+            )
+        },
+        // f(j, a, b, c) = (a AND b) OR (NOT(a) AND c)  (16 <= j <= 31)
+        // Note: It's the same as the sha256_ch function e.g. (a AND b) XOR (NOT(a) AND c)
+        // since the two logical expressions have the same truth table
+        round_idx if round_idx >= 16 && round_idx <= 31 => {
+            (
+                |a, b, c| (a & b) | ((!a) & c),
+                |cs, i, a, b, c| sha256_ch_boolean(cs.ns(|| format!("ch {}", i)), a, b, c)
+            )
+        },
+        // f(j, a, b, c) = (a OR NOT(b)) XOR c (32 <= j <= 47)
+        round_idx if round_idx >= 32 && round_idx <= 47 => {
+            (
+                |a, b, c| (a | (!b)) ^ c,
+                |cs, i, a, b, c| {
+                    let t = Boolean::or(cs.ns(|| format!("A OR NOT B {}", i)), &a, &b.not())?;
+                    Boolean::xor(cs.ns(|| format!("(A OR NOT B) XOR C {}", i)), &t, &c)
+                }
+            )
+        },
+        // f(j, a, b, c) = (c AND a) OR (NOT(c) AND b)  (48 <= j <= 63)
+        // Note: It's the same as the second round function, but with permuted variables
+        round_idx if round_idx >= 48 && round_idx <= 63 => {
+            (
+                |a, b, c| (c & a) | ((!c) & b),
+                |cs, i, a, b, c| sha256_ch_boolean(cs.ns(|| format!("permuted ch {}", i)), c, a, b)
+            )
+        },
+        // f(j, a, b, c) = (b OR NOT(c)) XOR a (64 <= j <= 79)
+        // Note: It's the same as the third round function but with permuted variables
+        round_idx if round_idx >= 64 && round_idx <= 79 => {
+            (
+                |a, b, c| (b | (!c)) ^ a,
+                |cs, i, a, b, c| {
+                    let t = Boolean::or(cs.ns(|| format!("B OR NOT C {}", i)), &b, &c.not())?;
+                    Boolean::xor(cs.ns(|| format!("(B OR NOT C) XOR a {}", i)), &t, &a)
+                }
+            )
+        },
+        _ => unreachable!()
+    };
+
+    triop(cs, a, b, c, tri_fn, circ_fn)
+}
+
+pub fn ripemd160_block_no_padding<ConstraintF, CS>(
+    mut cs: CS,
+    input: &[Boolean],
+) -> Result<Vec<Boolean>, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>,
+{
+    assert_eq!(input.len(), 512);
+
+    Ok(
+        ripemd160_compression_function(&mut cs, &input, &get_ripemd160_iv())?
+            .into_iter()
+            .flat_map(|e| e.to_bits_le())
+            .collect(),
+    )
+}
+
+pub fn ripemd160<ConstraintF, CS>(mut cs: CS, input: &[Boolean]) -> Result<Vec<Boolean>, SynthesisError>
+    where
+        ConstraintF: PrimeField,
+        CS: ConstraintSystem<ConstraintF>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    // append 0x80
+    padded.append(&mut UInt8::constant(1).into_bits_be());
```

Comment:
This part is a bit different from the sha256 (padded.push(Boolean::constant(true))). Why this?
Isn't the padding part the same?

Reply:
SHA256 assumes the input to be little endian while RIPEMD160 and Blake2S to be in Big Endian. That's why the difference